include/ginkgo/core/base/perturbation.hpp

```diff
@@ -0,0 +1,189 @@
+    friend class EnablePolymorphicObject<Perturbation, LinOp>;
+    friend class EnableCreateMethod<Perturbation>;
+
+public:
+    using value_type = ValueType;
+
+    /**
+     * Returns the basis of the perturbation.
+     *
+     * @return the basis of the perturbation
+     */
+    const std::shared_ptr<const LinOp> get_basis() const noexcept
+    {
+        return basis_;
+    }
+
+    /**
+     * Returns the projector of the perturbation.
+     *
+     * @return the projector of the perturbation
+     */
+    const std::shared_ptr<const LinOp> get_projector() const noexcept
+    {
+        return projector_;
+    }
+
+    /**
+     * Returns the scalar of the perturbation.
+     *
+     * @return the scalar of the perturbation
+     */
+    const std::shared_ptr<const LinOp> get_scalar() const noexcept
+    {
+        return scalar_;
+    }
+
+protected:
+    /**
+     * Creates an empty operator perturbation (0x0 operator).
+     *
+     * @param exec  Executor associated to the perturbation
+     */
+    explicit Perturbation(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Perturbation>(std::move(exec))
+    {}
+
+    /**
+     * Creates a perturbation with scalar and basis by setting projector to the
+     * conjugate transpose of basis. Basis must be transposable. Perturbation
+     * will throw gko::NotSupported if basis is not transposable.
+     *
+     * @param scalar  scaling of the movement
+     * @param basis  the direction basis
+     */
+    explicit Perturbation(std::shared_ptr<const LinOp> scalar,
+                          std::shared_ptr<const LinOp> basis)
+        : Perturbation(
+              std::move(scalar),
+              // basis can not be std::move(basis). Otherwise, Program deletes
+              // basis before applying conjugate transpose
+              basis,
+              std::move((as<gko::Transposable>(lend(basis)))->conj_transpose()))
+    {}
+
+    /**
+     * Creates a perturbation of scalar, basis and projector.
+     *
+     * @param scalar  scaling of the movement
+     * @param basis  the direction basis
+     * @param projector  decides the coefficient of basis
+     */
+    explicit Perturbation(std::shared_ptr<const LinOp> scalar,
+                          std::shared_ptr<const LinOp> basis,
+                          std::shared_ptr<const LinOp> projector)
+        : EnableLinOp<Perturbation>(basis->get_executor(),
+                                    gko::dim<2>{basis->get_size()[0]}),
+          scalar_{std::move(scalar)},
+          basis_{std::move(basis)},
+          projector_{std::move(projector)}
+    {
+        this->validate_perturbation();
+    }
+
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+    /**
+     * validate_perturbation check the dimension of scalar, basis, projector.
```

Comment:
No need to add the function name like this. I would suggest this: 

```suggestion
     * Validates the dimensions of the `scalar`, `basis` and `projector` parameters for the `apply`.
```